core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java

```diff
@@ -0,0 +1,246 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class TestUpdateRequirements {
+  private final TableMetadata metadata = mock(TableMetadata.class);
+
+  @BeforeEach
+  public void before() {
+    String uuid = UUID.randomUUID().toString();
+    when(metadata.uuid()).thenReturn(uuid);
+  }
+
+  @Test
+  public void nullCheck() {
+    assertThatThrownBy(() -> UpdateRequirements.forCreateTable(null))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage("Invalid metadata updates: null");
+
+    assertThatThrownBy(() -> UpdateRequirements.forUpdateTable(null, null))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage("Invalid table metadata: null");
+
+    assertThatThrownBy(() -> UpdateRequirements.forUpdateTable(metadata, null))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage("Invalid metadata updates: null");
+
+    assertThatThrownBy(() -> UpdateRequirements.forReplaceTable(null, null))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage("Invalid table metadata: null");
+
+    assertThatThrownBy(() -> UpdateRequirements.forReplaceTable(metadata, null))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage("Invalid metadata updates: null");
+  }
+
+  @Test
+  public void emptyUpdatesForCreateTable() {
+    assertThat(UpdateRequirements.forCreateTable(ImmutableList.of()))
+        .hasSize(1)
+        .hasOnlyElementsOfType(UpdateRequirement.AssertTableDoesNotExist.class);
+  }
+
+  @Test
+  public void emptyUpdatesForUpdateAndReplaceTable() {
+    assertThat(UpdateRequirements.forReplaceTable(metadata, ImmutableList.of()))
+        .hasSize(1)
+        .hasOnlyElementsOfType(UpdateRequirement.AssertTableUUID.class);
+
+    assertThat(UpdateRequirements.forUpdateTable(metadata, ImmutableList.of()))
+        .hasSize(1)
+        .hasOnlyElementsOfType(UpdateRequirement.AssertTableUUID.class);
+  }
+
+  @Test
+  public void assignUUID() {
+    assertThat(
+            UpdateRequirements.forUpdateTable(
+                metadata, ImmutableList.of(new MetadataUpdate.AssignUUID(metadata.uuid()))))
+        .hasSize(1)
+        .hasOnlyElementsOfType(UpdateRequirement.AssertTableUUID.class);
+  }
+
+  @Test
+  public void upgradeFormatVersion() {
+    assertThat(
+            UpdateRequirements.forUpdateTable(
+                metadata, ImmutableList.of(new MetadataUpdate.UpgradeFormatVersion(2))))
+        .hasSize(1)
+        .hasOnlyElementsOfType(UpdateRequirement.AssertTableUUID.class);
+  }
+
+  @Test
+  public void addSchema() {
+    assertThat(
+            UpdateRequirements.forUpdateTable(
+                metadata, ImmutableList.of(new MetadataUpdate.AddSchema(new Schema(), 0))))
+        .hasSize(2)
+        .hasOnlyElementsOfTypes(
+            UpdateRequirement.AssertTableUUID.class,
+            UpdateRequirement.AssertLastAssignedFieldId.class);
```

Comment:
I don't think that these tests are sufficient. The assertions must be based on the base metadata's state. So the AssertTableUUID must contain the base metadata's UUID. And the last assigned field ID assertion should contain the base metadata's last column ID. I think the implementation is correct, but this would not catch some bugs introduced by updating the implementation.

Reply:
makes sense and I've updated all tests to check values. I've also added more tests that check the failure conditions. Now we have full test coverage of all code branches in `UpdateRequirement` via `TestUpdateRequirements`